### PR TITLE
[Merged by Bors] - feat(Data/Finsupp/Defs): Support of `f + single a b`

### PR DESCRIPTION
**Pull request by vihdzp**
---

Used within my work with the Cantor normal form.

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/Finsupp/Defs.lean [hunk 1]
```diff
@@ -910,6 +910,20 @@ theorem support_add_eq [DecidableEq α] {g₁ g₂ : α →₀ M} (h : Disjoint
...
+theorem support_single_add [DecidableEq α] {a : α} {b : M} {f : α →₀ M}
+    (ha : a ∉ f.support) (hb : b ≠ 0) : support (single a b + f) = insert a f.support := by
+  have H := support_single_ne_zero a hb
+  apply (support_add_eq _).trans
+  · rw [H, insert_eq]
+  · rwa [H, disjoint_singleton_left]
+
+theorem support_add_single [DecidableEq α] {a : α} {b : M} {f : α →₀ M}
+    (ha : a ∉ f.support) (hb : b ≠ 0) : support (f + single a b) = insert a f.support := by
+  have H := support_single_ne_zero a hb
```
Comment (b-mehta) [inline Mathlib/Data/Finsupp/Defs.lean:922]: This one should have a shorter proof by dualising, right? (I'm guessing we don't have add_comm here)

Comment (vihdzp) [reply Mathlib/Data/Finsupp/Defs.lean:922]: Yes, this works even without commutative addition. Is there any type alias for reversing the order of addition?

**Comment by b-mehta** [comment]
aod, I think. @YaelDillies will know

File: Mathlib/Data/Finsupp/Defs.lean (hunk 1 above)
Comment (YaelDillies) [reply Mathlib/Data/Finsupp/Defs.lean:922]: `AddOpposite`

**Review by Ruben-VandeVelde — APPROVED** [review]
Seems like deduplicating the proofs is non-obvious, so

maintainer merge

File: Mathlib/Data/Finsupp/Defs.lean [hunk 2]
```diff
@@ -910,6 +910,20 @@ theorem support_add_eq [DecidableEq α] {g₁ g₂ : α →₀ M} (h : Disjoint
...
+  apply (support_add_eq _).trans
+  · rw [H, union_comm, insert_eq]
+  · rwa [H, disjoint_singleton_right]
```
Comment (eric-wieser) [inline Mathlib/Data/Finsupp/Defs.lean:925]: Perhaps better as
```suggestion
theorem support_add_single  {a : α} {b : M} {f : α →₀ M}
    (ha : a ∉ f.support) (hb : b ≠ 0) : support (f + single a b) = cons a f.support ha := by
  classical
  have H := support_single_ne_zero a hb
  rw [support_add_eq, H, union_comm, cons_eq_insert, insert_eq]
  rwa [H, disjoint_singleton_right]
```

**Review by Ruben-VandeVelde — APPROVED** [review]
Thanks!

maintainer merge

**Review by eric-wieser — APPROVED** [review]
bors merge

Thanks!

**Pull request closed without merging**